include/dali/core/per_stream_pool.h

```diff
@@ -0,0 +1,206 @@
+
+    template <typename... ObjectConstructorArgs>
+    ListNode(int device_id, cudaStream_t stream, ObjectConstructorArgs&&... args)
+    : object(std::forward<ObjectConstructorArgs>(args)...)
+    , device_id(device_id)
+    , stream(stream)
+    , event(CUDAEvent::Create()) {
+    }
+
+    ~ListNode() {
+      DeleteNonRecursive(std::move(next));
+    }
+
+    static void DeleteNonRecursive(ListNodeUPtr ptr) {
+      while (ptr) {
+        auto tmp = std::move(ptr->next);
+        // ptr will be deleted, but its next pointer is null now, so we'll not recurse
+        ptr = std::move(tmp);
+      }
+    }
+  };
+
+ public:
+  class ObjectLease {
+   public:
+    ObjectLease(ObjectLease &&) = default;
+    ~ObjectLease() { owner->Release(std::move(node)); }
+    operator T*() const { return &node->object; }
+    T &operator*() const { return node->object; }
+    T *operator->() const { return &node->object; }
+    explicit operator bool() const noexcept { return node != nullptr; }
+   private:
+    ObjectLease(PerStreamPool *owner, ListNodeUPtr node) : owner(owner), node(std::move(node)) {}
+    PerStreamPool *owner;
+    ListNodeUPtr node;
+    friend class PerStreamPool;
+  };
+
+  ObjectLease Get(int device_id, cudaStream_t stream) {
+    std::lock_guard<mutex_type> guard(lock_);
+    RecycleFinished();
+    if (reuse_pending_on_same_stream) {
+      if (auto ret = GetPending(stream))
+          return { this, std::move(ret) };
+    }
+    return { this, GetFromDevicePool(device_id, stream) };
+  }
+
+  ObjectLease Get(cudaStream_t stream) {
+    int device = -1;
+    cudaGetDevice(&device);
+    return Get(device, stream);
+  }
+
+ private:
+  friend class ObjectLease;
+
+  void Release(std::unique_ptr<ListNode> node) {
+    cudaEventRecord(node->event, node->stream);
+    std::lock_guard<mutex_type> guard(lock_);
+    node->next = std::move(pending_);
+    pending_ = std::move(node);
+  }
+
+  ListNodeUPtr GetPending(cudaStream_t stream) {
+    for (ListNodeUPtr *pptr = &pending_; *pptr; pptr = &(*pptr)->next) {
+      if ((*pptr)->stream == stream) {
+        ListNodeUPtr ret = std::move(*pptr);
+        *pptr = std::move(ret->next);
+        return ret;
+      }
+    }
+    return nullptr;
+  }
+
+  ListNodeUPtr GetFromDevicePool(int device_id, cudaStream_t stream) {
+    int idx = device_id + 1;  // account for device -1
+    if (idx >= static_cast<int>(dev_pools_.size()) || !dev_pools_[idx])
+      return std::make_unique<ListNode>(device_id, stream);
+
+    ListNodeUPtr ptr = std::move(dev_pools_[idx]);
+    dev_pools_[idx] = std::move(ptr->next);
+    return ptr;
+  }
+
+  void RecycleFinished() {
+    for (ListNodeUPtr *pptr = &pending_; *pptr; ) {
+      if (cudaErrorNotReady == cudaEventQuery((*pptr)->event)) {  // still pending_? skip!
+        pptr = &(*pptr)->next;
+      } else {  // otherwise it's finished (or an error; we still recycle)
+        int idx = (*pptr)->device_id + 1;
+        if (idx >= dev_pools_.size())
```

Comment:
warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
         if (idx >= dev_pools_.size())